### ISSUE TITLE
State out of date when switching routes

### ISSUE THREAD

**Issue by jondashkyle** [choojs/choo#530]
### Expected behavior
When attaching a callback to `state.events.NAVIGATE`,  and navigating to a new route, I expect `state.route` to represent the current route.

### Actual behavior
Instead, `state.route` is out of date. It should displaying the current route, but instead shows the previous. For example, if navigating from a route called `/cute` to one called `/nice`, it will return `/cute`.

```js
function model (state, emitter) {
  emitter.on(state.events.NAVIGATE, function () {
     console.log(state.route)
  })
}
```

**Comment by yoshuawuyts** [comment]
Good catch! - think the fix for this would be to update `state.route`
whenever we call navigate

**Comment by yoshuawuyts** [comment]
_ah!_ ok so this is tricky, because we can't know the route _until after_ we navigate. Why? Because it's the router that figures out which route we'll be using to render. The `'navigate'` event only triggers the router.

So the solution for this is to do:
```js
function model (state, emitter) {
  emitter.prependListener(state.events.NAVIGATE, function () {
    emitter.once(state.events.RENDER, function () {
       console.log(state.route)
    })
  })
}
```

The flow will then be:
```txt
navigate triggered -> navigate event fired -> model triggered ->
model render event attached (once) -> render event fired ->
router triggered -> new route set on state -> render complete ->
model render event triggered
```

Does this make sense? Hope it does haha; wish we could simplify it, but not sure how we could do this (besides, like, writing a module). Thanks!

**Comment by akiva** [comment]
This is the same issue I am encountering in #538 , using `choo@next`, and the suggested fix doesn't actually work.

**Comment by akiva** [comment]
OK, while the suggested fix didn't work for me, _this_ does:

```javascript
const app = require('choo')()
const html = require('choo/html')

app.use(function (state, emitter) {
  state.loggedIn = false

  emitter.on('DOMContentLoaded', function () {
    if (state.route !== '/shazam' && !state.loggedIn) {
      emitter.emit('pushState', '/shazam')
    }
  })

  emitter.prependListener('DOMContentLoaded', function () {
    if (state.route !== '/shazam' && !state.loggedIn) {
      emitter.emit('pushState', '/shazam')
      emitter.emit('render')
    }
  })

  emitter.prependListener('navigate', function () {
    console.log('state.route:', state.route, window.location.pathname)
    state.route = window.location.pathname
    emitter.once('render', function () {
      console.log('state.route:', state.route, window.location.pathname)
    })
  })

  emitter.on('navigate', function () {
    console.log('Post-navigate event route', state.route)
  })
})

app.route('/', function (state, emit) {
  return html`
    <body>
      <p>Foo bar baz quux</p>
    </body>
  `
})

app.route('/shazam', function (state, emit) {
  return html`
    <body>
      <p>Shazam</p>
    </body>
  `
})

if (module.parent) module.exports = app
else app.mount('body')
```

@yoshuawuyts Am I doing something wrong here? I would be happy to work on a proper fix to merge into `choo`.

**Comment by akiva** [comment]
OK, while the suggested fix didn't work for me, _this_ does:

```javascript
const app = require('choo')()
const html = require('choo/html')

app.use(function (state, emitter) {
  state.loggedIn = false

  emitter.prependListener('DOMContentLoaded', function () {
    if (state.route !== '/shazam' && !state.loggedIn) {
      emitter.emit('pushState', '/shazam')
      emitter.emit('render')
    }
  })

  emitter.prependListener('navigate', function () {
    console.log('state.route:', state.route, window.location.pathname)
    state.route = window.location.pathname
    emitter.once('render', function () {
      console.log('state.route:', state.route, window.location.pathname)
    })
  })

  emitter.on('navigate', function () {
    console.log('Post-navigate event route', state.route)
  })
})

app.route('/', function (state, emit) {
  return html`
    <body>
      <p>Foo bar baz quux</p>
    </body>
  `
})

app.route('/shazam', function (state, emit) {
  return html`
    <body>
      <p>Shazam</p>
    </body>
  `
})

if (module.parent) module.exports = app
else app.mount('body')
```

@yoshuawuyts Am I doing something wrong here? I would be happy to work on a proper fix to merge into `choo`.

**Comment by yoshuawuyts** [comment]
@akiva o/ whoops, missed your reply. Don't think that works, because the `route` is the literal route that's resolved from the browser. What you're setting there should be available from `state.href` too!

The only way we could get the route would be to attach a listener that resolves after the router has resolved. Think that could be done by waiting for the `DOMContentLoaded` event tbh ( ref: https://github.com/choojs/choo/blob/master/index.js#L85-L96). Ok, try this:

```js
function store (state, emitter) {
  emitter.on(state.events.DOMCONTENTLOADED, function () {
    emitter.on(state.events.NAVIGATE, function () {
       console.log(state.route)
    })
  })
}
```

**Comment by jongacnik** [comment]
I've not yet looked into why, and perhaps it's a misunderstanding of the suggested flow above, but I have found that to get this to work 100%, `RENDER` needs to be explicitly triggered within the `NAVIGATE` callback (def a hack!). I've added examples, view console to see behaviors.

#### Explicit Render

The `RENDER` event is picked up after `NAVIGATE` because we force a re-render

```js
emitter.prependListener(state.events.NAVIGATE, function () {
  setTimeout(function () {
    emitter.emit(state.events.RENDER)
  })
  emitter.once(state.events.RENDER, function () {
    console.log(state.route)
  })
})
``` 

Demo: https://choo-navigate-render-hack.glitch.me
Code: https://glitch.com/edit/#!/choo-navigate-render-hack

#### Without Explicit Render

The `RENDER` event does not fire until the next tick

```js
emitter.prependListener(state.events.NAVIGATE, function () {
  emitter.once(state.events.RENDER, function () {
    console.log(state.route)
  })
})
```

Demo: https://choo-navigate-render.glitch.me
Code: https://glitch.com/edit/#!/choo-navigate-render

**Comment by yoshuawuyts** [comment]
Should be fixed in `choo@6.7.0` :tada:
